util/src/trie/triedb.rs

```diff
@@ -127,18 +128,19 @@ impl<'db> TrieDB<'db> {
 	}
 
 	/// Get the root node's RLP.
-	fn root_node(&self) -> super::Result<Node> {
-		self.root_data().map(Node::decoded)
+	fn root_node<'a, R: 'a + Recorder>(&self, r: &'a mut R) -> super::Result<Node> {
+		self.root_data(r).map(Node::decoded)
 	}
 
 	/// Get the data of the root node.
-	fn root_data(&self) -> super::Result<&[u8]> {
+	fn root_data<'a, R: 'a + Recorder>(&self, r: &'a mut R) -> super::Result<&[u8]> {
 		self.db.get(self.root).ok_or_else(|| Box::new(TrieError::InvalidStateRoot(*self.root)))
+			.map(|node| { r.record(self.root, node, 0); node })
 	}
 
 	/// Get the root node as a `Node`.
-	fn get_node(&'db self, node: &'db [u8]) -> super::Result<Node> {
-		self.get_raw_or_lookup(node).map(Node::decoded)
+	fn get_node<'a, R: 'a + Recorder>(&'db self, node: &'db [u8], r: &'a mut R, depth: u32) -> super::Result<Node> {
```

Comment:
`LES` can require you to omit nodes below a certain depth. i can imagine a scenario where you keep the first `N` layers of the CHT and falls back to the network when that fails.